src/hotspot/share/c1/c1_Instruction.cpp

```diff
@@ -138,15 +138,13 @@ bool Instruction::maybe_flattened_array() {
   if (UseFlatArray) {
     ciType* type = declared_type();
     if (type != NULL) {
-      if (type->is_obj_array_klass()) {
-        // Due to array covariance, the runtime type might be a flattened array.
+      if (type->is_obj_array_klass() && !type->as_obj_array_klass()->is_elem_null_free()) {
+        // The runtime type of [LMyValue might be [QMyValue due to [QMyValue <: [LMyValue.
```

Comment:
if type->is_obj_array_klass() is true and type->as_obj_array_klass()->is_elem_null_free() is also true, the method returns false, is it the expected behavior?

Reply:
Yes, because in that case we have a null-free but non-flattened array. Right?

Reply:
Correct. Got it.